bam_fastq.c

```diff
@@ -880,135 +570,287 @@ static inline bool filter_it_out(const bam1_t *b, const bam2fq_state_t *state)
+int output_index(bam1_t *b1, bam1_t *b2, bam2fq_state_t *state,
+                 bam2fq_opts_t* opts) {
+    bam1_t *b[2] = {b1, b2};
+
+    char *ifmt = opts->index_format;
+    if (!ifmt)
+        ifmt = "i*i*";
+
+    // Get seq / qual elements
+    char *bc = (char *)bam_aux_get(b1, opts->barcode_tag);
+    if (!bc)
+        bc = (char *)bam_aux_get(b2, opts->barcode_tag);
+    if (!bc)
+        return 0;
+    else
+        bc++; // skip Z
+
+    char *qt = (char *)bam_aux_get(b1, opts->quality_tag);
+    if (!qt)
+        qt = (char *)bam_aux_get(b2, opts->quality_tag);
+    if (qt && strlen(bc) != strlen(qt)-1)
+        qt = NULL;
+    else if (qt)
+        qt++;
+
+    int inum = 0;
+    while (inum < 2) {
+        char fc = *ifmt++;
+        if (!fc)
+            break; // ran out of index-format
+
+        long len, rem = 0;
+        if (isdigit(*ifmt)) {
+            rem = len = strtol(ifmt, &ifmt, 10);
+        } else {
+            ifmt++;
+            len = 0;
+        }
+
+        char *bc_end = bc, *qt_end = qt;
+        while (len ? *bc_end && rem-- : isalpha(*bc_end))
+            bc_end++, qt_end += qt != NULL;
+
+        switch (fc) {
+        case 'n':
+            // skip
+            bc = bc_end + (len==0);
+            if (qt)
+                qt = qt_end + (len==0);
+            break;
+
+        case 'i':
+            if (write_index_rec(state->fpi[inum], b[inum], state, opts,
+                                bc, bc_end-bc, qt, qt_end-qt) < 0)
+                return -1;
+            bc = bc_end + (len==0);
+            if (qt)
+                qt = qt_end + (len==0);
+            inum++;
+            break;
+
+        default:
+            fprintf(stderr, "Unknown index-format code, skipping\n");
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+static int flush_rec(bam2fq_state_t *state, bam2fq_opts_t* opts,
+                     bam1_t *b[4], int score[3], int best[3],
+                     int64_t *n_singletons) {
+    if (score[1] > 0 && score[2] > 0) {
+        // If CASAVA tag is required and barcode is only on R1,
+        // copy it to R2
+        if (state->illumina_tag) {
+            char *tag;
+            if ((tag = (char *)bam_aux_get(b[best[1]],
+                                           opts->barcode_tag)))
+                bam_aux_update_str(b[best[2]],
+                                   opts->barcode_tag,
+                                   strlen(tag), tag+1);
+            if ((tag = (char *)bam_aux_get(b[best[1]],
+                                           opts->quality_tag)))
+                bam_aux_update_str(b[best[2]],
+                                   opts->quality_tag,
+                                   strlen(tag), tag+1);
+
+        }
+        if (sam_write1(state->fpr[1], state->h, b[best[1]]) < 0)
+            goto err;
+        if (sam_write1(state->fpr[2], state->h, b[best[2]]) < 0)
+            goto err;
+
+        output_index(b[best[1]], b[best[2]], state, opts);
```

Comment:
It's far too long ago for me to remember whether it's deliberate or not!  However I was working off the Illumina CASAVA documentation at the time.

Do we know if single-ended CASAVA tagged fastq is a real thing?

Reply:
I don't know about single-ended CASAVA, but I don't see why it wouldn't happen if you ran a single-ended run for some reason.  Single-ended barcoded data is very much a thing (e.g. PacBio).

Reply:
> Also, the CASAVA and index outputs only work for paired reads at the moment. Is this deliberate? It would cause problems if you have barcoded single-ended reads.

Can you give an example please?  It works for me.

Eg:

```
@ [samtools.../samtools]; samtools import -i -s test/import/3.interleaved.fq  | samtools view -f 0x40 | head -2
ref1_grp1_p001	77	*	0	0	*	*	0	0	CGAGCTCGGT	!!!!!!!!!!	BC:Z:AAA+CCC
ref1_grp1_p002	77	*	0	0	*	*	0	0	CTCGGTACCC	##########	BC:Z:AAA+CCC
@ [samtools.../samtools]; samtools import -i -s test/import/3.interleaved.fq  | samtools view -f 0x40 | head -2 | samtools fastq -i --index-format 'i*i*' 2>/dev/null
@ref1_grp1_p001/1 1:N:0:AAA+CCC
CGAGCTCGGT
+
!!!!!!!!!!
@ref1_grp1_p002/1 1:N:0:AAA+CCC
CTCGGTACCC
+
##########
@ [samtools.../samtools]; samtools import -i -s test/import/3.interleaved.fq  | samtools view -f 0x40 | head -2 | sed 's/77/4/' | samtools fastq -i --index-format 'i*i*' 2>/dev/null
@ref1_grp1_p001 0:N:0:AAA+CCC
CGAGCTCGGT
+
!!!!!!!!!!
@ref1_grp1_p002 0:N:0:AAA+CCC
CTCGGTACCC
+
##########
```

Hacking the SAM flag from 77 (PAIRED,UNMAP,MUNMAP,READ1) to 4 (UNMAP) doesn't change whether CASAVA tags can be output, although it does change the read number in the casava output (1:N:0 to 0:N:0) which I believe to be correct.

Reply:
Sorry, I was mistaken about the CASAVA tag - it is written out as you say.  But if you use the `--i1` and `--i2` as well then nothing gets written to them as `output_index()` is only called for paired reads.  This is a change from the old version which wrote the barcode data files before choosing which output file(s) to put the read data in, so you would also get barcode data for single-ended reads.

The [bcl2fastq documentation](https://support.illumina.com/content/dam/illumina-support/documents/documentation/software_documentation/bcl2fastq/bcl2fastq_letterbooklet_15038058brpmi.pdf) says read number "1 can be single read ..." so I think they should be "1:N:0" - although it then says "... or read 2 of paired-end" which is a bit confusing (possibly they meant read 1?)

Reply:
Regarding `0:N:0` vs `1:N:0` that's not something I changed.  See https://github.com/samtools/samtools/blob/develop/bam_fastq.c#L290-L292 for the original code.

I don't have access to the Illumina software nor any raw BCL data, so I've no idea how it operates in practice.  In absence of that I just assumed that the original had been written with the correct output.   I'd want some actual data if we were to change it to verify the old code was bugged.

The --i1 and --i2 thing I'll look into.

Reply:
The `--i1` and `--i2` options are working better, but I'm still not getting them for single-ended runs.  For an example. try:
```
wget 'http://ftp.sra.ebi.ac.uk/vol1/run/ERR194/ERR1940805/21943_1%2311.cram'
samtools fastq -i -1 1.fq -2 2.fq -0 0.fq -s s.fq --i1 /tmp/i1.fq --index-format 'i*' 21943_1#11.cram
wc -l *.fq
  9917212 0.fq
        0 1.fq
        0 2.fq
        0 i1.fq
        0 s.fq
  9917212 total
```

Reply:
A good job you chose a file we also have locally!  ENA once again on a < 100KB/s transfer rate.  Argh.

I see this is actually the case which was clearly something I was scratching my head over when writing this code (ie trying to understand what data it pertained to probably as the original algorithm wasn't obvious):

```
if (score[0]) { // TODO: check this
```

Looks easy to fix fortunately. Thanks for the clear example.